src/jit/hwintrinsicxarch.cpp

```diff
@@ -1044,6 +1045,116 @@ GenTree* Compiler::impSSE2Intrinsic(NamedIntrinsic        intrinsic,
             break;
         }
 
+        case NI_SSE2_SetAllVector128:
+        {
+            assert(sig->numArgs == 1);
+            baseType = getBaseTypeOfSIMDType(sig->retTypeSigClass);
+            assert((baseType >= TYP_BYTE) && (baseType <= TYP_DOUBLE));
+            assert(baseType != TYP_FLOAT);
+
+            op1          = impPopStack().val;
+            GenTree* src = nullptr;
+
+            switch (baseType)
+            {
+                case TYP_DOUBLE:
+                {
+                    src     = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, NI_SSE2_SetScalarVector128, baseType, simdSize);
+                    retNode = gtNewSimdHWIntrinsicNode(TYP_SIMD16, src, gtNewIconNode(0b01000100), NI_SSE2_Shuffle,
```

Comment:
In the future, we may need a phase to eliminate "floating conversion" intrinsics (like compiler-generated or user-written `NI_SSE2_SetScalarVector128`).
In most cases, these intrinsics just make the type system happy (match `Vector128<float/double>` with `float/double`), but the codegen is unnecessary.

Reply:
Can we directly change the return type of `op1` to `TYP_SIMD16` to avoid `NI_SSE2_SetScalarVector128` here? Or make a new IR to match the types.
The current CQ of  `NI_SSE/SSE2_SetScalarVector128` is not good due to the "zero upper" semantics.

@CarolEidt @AndyAyersMS

Reply:
Actually, since CPU instructions are type agnostic we should have a type free but only operand size constrained internal codegen for intrinsics.

Reply:
> In the future, we may need a phase to eliminate "floating conversion" intrinsics (like compiler-generated or user-written NI_SSE2_SetScalarVector128).

Reply:
@mikedn Looks a new helper-intrinsic is simpler? https://github.com/dotnet/coreclr/pull/16772

Reply:
> Can we directly change the return type of op1 to TYP_SIMD16 to avoid NI_SSE2_SetScalarVector128 here?

I don't think that is possible without introducing non-deterministic behavior.

---
In reply to: [172425955](https://github.com/dotnet/coreclr/pull/16758#discussion_r172425955) [](ancestors = 172425955)